README.md
# Marple Python Modules

A python package for modules shared across the Marple project.


### Get started

`make install` (sets up local environment and installs dependencies)

### Install package

`pip install git+ssh://git@github.com/marple-newsrobot/marple-py-modules.git`

...or, using https:

`pip install git+https://github.com/marple-newsrobot/marple-py-modules.git`

### Run tests

`make tests` (run all tests)

`make test file=path/to/test` (run specific test)

### Deploy

`make deploy_new_version v=0.0.4 msg="Made some changes"`

The current version is defined in `CURRENT_VERSION.txt`. This file is updated with this make command.

### Update docs

`cd docs`

`make html`

### Notes to self

Useful links:

- Private packages, versioning etc. http://jtushman.github.io/blog/2013/06/17/sharing-code-across-applications-with-python/#3
- Generate documentation with Sphinx: http://gisellezeno.com/tutorials/sphinx-for-python-documentation.html



CHANGES.txt
v0.0.1, 2016-11-21 -- First version
v0.0.2, 2016-11-27 -- Robot reporter compatible version
v0.0.3, 2016-12-02 -- Added deploy scripts
v0.0.4, 2016-12-02 -- Enable querying with lists
v0.0.5, 2016-12-07 -- Added method .children() to Domain
v0.0.6, 2016-12-12 -- Added AWS connection
v0.0.7, 2017-01-30 -- Added notes to dataset
v0.0.8, 2017-02-01 -- Added notes_from_csv method to dataset
v0.0.9, 2017-02-01 -- Added update option to append method
v0.0.10, 2017-02-13 -- Bug fixes
v0.0.11, 2017-02-16 -- Simplify schema structure (forces restructuring of marple-datasets/schema)
v0.0.11, 2017-02-16 -- Simplify schema structure (forces restructuring of marple-datasets/schema)
v0.0.12, 2017-02-17 -- Refactor CsvFile
v0.0.13, 2017-03-03 -- Added multilanguage support to schema
v0.0.14, 2017-03-21 -- Added recipe connection
v0.0.15, 2017-03-27 -- Add delete method to db connection + base class for worker
v0.0.16, 2017-04-05 -- Added gt, gte, lt and lte methods to postgrest
v0.0.17, 2017-04-06 -- Handle invalid query on local connection
v0.0.18, 2017-04-13 -- Get recipe by id from db properly
v0.0.19, 2017-04-20 -- Add pipeline connection
v0.0.20, 2017-04-20 -- Add region to as connection
v0.0.21, 2017-05-09 -- Add function for subtracting periods from timepoint to utils
v0.0.22, 2017-06-20 -- Use glob2 and fix label method
v0.0.23, 2017-07-01 -- correct required column in datasets.csv
v0.0.24, 2017-07-05 -- reintroduce datetype class
v0.0.25, 2017-07-05 -- fix datatypes
v0.0.26, 2017-07-25 -- add multilingual label property to datatypes
v0.0.27, 2017-07-25 -- actually add multilingual label property to datatypes
v0.0.28, 2017-08-07 -- Enable caching on requests
v0.0.29, 2017-08-07 -- Correct dependency
v0.0.29, 2017-08-07 -- Correct dependency
v0.0.30, 2017-08-14 -- better .exists() method
v0.0.31, 2017-08-14 -- enable cache param on get_by_id on local connection
v0.0.32, 2017-11-09 -- add __repr__ and .all_nots
v0.0.33, 2017-11-14 -- Handle units + __repr__ > __unicode__
v0.0.33, 2017-11-14 -- Handle units + __repr__ > __unicode__
v0.0.34, 2017-11-16 -- handle quarterly data
v0.0.35, 2017-11-21 -- richer queries with children method
v0.0.36, 2017-12-20 -- Handle weekly data
v0.0.37, 2018-03-19 -- handle school year
v0.0.38, 2018-08-02 -- update dependency version
v0.0.39, 2018-09-03 -- trying to silence some warnings
v0.0.40, 2018-12-14 -- add .get_by_id to aws connection
v0.0.41, 2018-12-14 -- same as prev, but with actual commits
v0.0.42, 2018-12-14 -- revert to old requirements
v0.0.43, 2018-12-20 -- handle restricted datasets
v0.0.44, 2019-01-03 -- handle stricter auth on datasets
v0.0.45, 2019-01-09 -- update to handle database restrictions
v0.0.46, 2019-01-14 -- Significantly reduce number of dependencies, and upgrade the remaining ones where possible
v0.0.47, 2019-01-14 -- Fix setup error
v0.0.48, 2019-08-12 -- decimal precision on dataset to json
v0.0.49, 2019-08-12 -- adds to_json (string) method
v0.0.50, 2019-08-29 -- first attempt att python3 compability
v0.0.51, 2019-08-30 -- handle missing values
v0.0.52, 2019-09-19 -- add connection for tabular data
v0.0.53, 2019-09-19 -- make sure upsert returns something
v0.0.54, 2019-09-19 -- better error message on connection error
v0.0.55, 2019-09-30 -- add custom method to postgrest api
v0.0.56, 2019-09-30 -- bug fix: enable chaning
v0.1.0, 2019-10-08 -- move connections to separate package
v0.1.1, 2019-10-08 -- move connections to separate package
v0.1.17, 2019-11-20 -- fix is_na bug
v0.1.18, 2019-11-20 -- handle integers in parse_decimal
v0.1.19, 2021-01-14 -- new version of nw connections
v0.1.19, 2021-01-14 -- nw connections 1.1.1
v0.1.20, 2021-01-28 -- enable higher versions of jsonschema, pandas and numpy
v0.1.21, 2021-01-28 -- enable higher versions of jsonschema, pandas and numpy in setup.pyas well
v0.1.22, 2021-03-16 -- Remove csvkit dependency
v0.1.23, 2021-03-16 -- Remove csvkit dependency - from setup.py as well
v0.2.0, 2021-09-19 -- Bump nw-connections
v0.2.1, 2021-09-19 -- Bump nw-connections
v0.2.14, 2021-09-30 -- bug fix: handle nan in pandas>1.2.0
v0.2.15, 2021-11-18 -- update nw connections
v0.2.16, 2023-04-06 -- Use concat instead of append in pandas
v0.2.17, 2023-06-19 -- Fix .drop() bug in.


marple/utils.py
# encoding: utf-8

from datetime import datetime
from dateutil.relativedelta import relativedelta
import os
import re
import requests_cache
import functools
import json
from decimal import Decimal
import numpy as np
import pandas as pd
from six import string_types, text_type

def get_timepoint_label(datestring, periodicity):
    """ Convert a datestring to a timepoint label.
        :param datestring: an iso coded datestring. E.g. "2016-01-01"
        :param periodicity: monthly|quarterly|yearly|rolling_quarter|rolling_year
        :returns: a label for the timepoint
        :rtype: str
    """
    timepoint = datetime.strptime(datestring, '%Y-%m-%d')

    if periodicity == "monthly":
        # Jan 2016
        return timepoint.strftime("%b %Y")

    elif periodicity == "quarterly":
        # Q1 2016
        quarters = [0, 1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 4, 4]
        return "Q{} {}".format(
            quarters[timepoint.month],
            timepoint.year)

    elif periodicity == "yearly":
        # 2016
        return str(timepoint.year)

    elif periodicity == "weekly":
        return "week {}".format(timepoint.week)

    elif periodicity == "rolling_quarter":
        # 2016-03-01 => "Jan 2016-Mar 2016"
        timepoint_start = timepoint - relativedelta(months=2)
        return "{}-{}".format(
            timepoint_start.strftime("%b %Y"),
            timepoint.strftime("%b %Y"))

    elif periodicity == "rolling_year":
        # 2017-01-01 => "Feb 2016-Jan 2017"
        timepoint_start = timepoint - relativedelta(months=11)
        return "{}-{}".format(
            timepoint_start.strftime("%b %Y"),
            timepoint.strftime("%b %Y"),
        )

    raise Exception(u"Unknown periodicity: '{}'".format(periodicity))

REGEX = {
    "yearly": "^\d\d\d\d$",
    "quarterly": "^\d\d\d\d-?[KQ][1-4]$",
    "monthly": "^\d\d\d\d-\d\d$",
}
def guess_periodicity(time_str):
    """ Guess periodicity from a date string
        "2015" => "yearly"
        "2015Q1" => "quarterly"
        "2015-K1" => "quarterly"
        "2015-01" => "monthly"
    """
    time_str = text_type(time_str)
    if re.match(REGEX["yearly"], time_str):
        return "yearly"
    elif re.match(REGEX["quarterly"], time_str):
            return "quarterly"
    elif re.match(REGEX["monthly"], time_str):
            return "monthly"
    else:
        raise ValueError(u"Unknown time string: '{}'".format(time_str))


def to_timepoint(time_str):
    """ Get the starting timepoint of period
        "2015" => "2015-01-01"
        "2015-03" => "2015-03-01"
    """
    time_str = text_type(time_str)
    if re.match("^\d\d\d\d$", time_str):
        # "2015"
        return time_str + "-01-01"
    elif re.match(REGEX["monthly"], time_str):
        # "2015-01"
        return time_str + "-01"
    elif re.match("^\d\d\d\d-\d\d-\d\d$", time_str):
        # 2015-01-01
        return time_str
    elif re.match(REGEX["quarterly"], time_str):
        year = time_str[:4]
        quarter_i = parse_int(time_str[4:].replace("K","").replace("Q","").replace("-",""))
        month = quarter_i * 3 - 2
        return "{}-{:02d}-01".format(year, month)
    else:
        raise ValueError(u"Unknown time string: '{}'".format(time_str))



def subtract_periods(timepoint, n_periods, periodicity=None):
    """ Subtract n number of periods from a timepoint
    Example usage:

        subtract_periods(2015, 2, "yearly") => "2013-01-01"
        subtract_periods("2015-03", 2, "monthly") => "2015-01-01"
        subtract_periods("2015-07-01", 6, "monthly") => "2015-01-01"

    :param timepoint (str|int): a timepoint
    :param n_periods (str): number of periods
    :param periodicity (str): "monthly"|"yearly"|"rolling_quarter"|"rolling_year"
    :returns: computed timepoint as string
    """
    if periodicity is None:
        periodicity = guess_periodicity(timepoint)

    timepoint = to_timepoint(timepoint)
    dt = datetime.strptime(timepoint, "%Y-%m-%d")

    if periodicity in ["monthly", "rolling_quarter", "rolling_year"]:
        dt -= relativedelta(months=n_periods)
    elif periodicity in ["yearly", "school_year"]:
        dt -= relativedelta(years=n_periods)
    else:
        msg = u"Unable to subtract periods for periodicity '{}'".format(periodicity)
        NotImplementedError(msg)

    return datetime.strftime(dt, "%Y-%m-%d")


def list_files(dir, extension=None, file_name=None):
    """ Get a list of all files in directory and subdirectories
        :param dir: path to direcotyr to parse
        :param extension: Only include files with given extension.
            Both ".txt" and "txt" allowed.
    """
    r = []
    subdirs = [x[0] for x in os.walk(dir)]
    for subdir in subdirs:
        files = next(os.walk(subdir))[2]
        if (len(files) > 0):
            for file in files:
                _file_name, _file_extension = os.path.splitext(file)

                if extension != None:
                    _file_extension = _file_extension.replace(".","")
                    extension = extension.replace(".","")

                    if extension != _file_extension:
                        continue

                if file_name != None:
                    if file_name != file:
                        continue

                r.append(subdir + "/" + file)

    return r

def parse_lingual_object(str_or_dict, lang=None, prefix=None, fallback_chain=["en","sv"]):
    """ Json objects my contain strings in multiple languages.
        This

        Eg.
        parse_lingual_object({"en": "dog", "sv": "hund"}, "en") => "dog"
        parse_lingual_object({"en": "dog", "sv": "hund"}, "sv") => "hund"
        parse_lingual_object({"en": "dog", "sv": "hund"}, None) => "dog"
        parse_lingual_object({
            "label__en": "dog",
            "label__sv": "hund"}, "sv", "label") => "hund"
        parse_lingual_object("dog") => "dog"

        :param str_or_dict: string or dict to parse
        :param prefix: ie. "label" or "description" if key name is "label__en"
        :param fallback_chain: In case no language is defined, pick
            one of the fallback langagues.
        :returns: a translated string
    """
    #
    x = str_or_dict

    # 1. If string, return as such
    if isinstance(x, string_types):
        return x

    #if lang is None and prefix is not None:
    #    return x[prefix]

    # 2. Parse dict
    elif isinstance(x, dict):
        # 2.1 Parse using defined languagre
        if lang is not None:
            try:
                if prefix is not None:
                    key = "{}__{}".format(prefix, lang)
                else:
                    key = lang
                return x[key]
            except KeyError:
                raise KeyError("Unable to translate '{}' with '{}'"\
                    .format(x, key))


        else:
            # 2.2 Check if there is a default ie "label" (without any
            # language key)
            if prefix is not None and prefix in x:
                return x[prefix]

            # 2.3 Try fallback languages
            for fallback_lang in fallback_chain:
                if prefix is not None:
                    key = "{}__{}".format(prefix, fallback_lang)
                else:
                    key = fallback_lang
                try:
                    return x[key]
                except KeyError:
                    pass

    raise ValueError("Unable to translate '{}' with '{}'"\
            .format(x,lang))

def parse_int(s):
    return int(float(s))


def parse_decimal(val):
    if val is None or pd.isna(val):
        return None
    elif isinstance(val, float) or isinstance(val, int):
        return Decimal(val)
    else:
        raise NotImplementedError(u"Unable to parse Decimal from {}".format(val))

def isNaN(num):
    """ Check if a value is nan
        http://stackoverflow.com/questions/944700/how-to-check-for-nan-in-python
    """
    return num != num

def multiple_and(*conditions):
    """Pass multiple conditions to

    >>> c_1 = data.col1 == True
    >>> c_2 = data.col2 < 64
    >>> c_3 = data.col3 != 4
    >>> data_filtered = data[multiple_and(c1,c2,c3)]
    https://stackoverflow.com/questions/13611065/efficient-way-to-apply-multiple-filters-to-pandas-dataframe-or-series
    """
    return functools.reduce(np.logical_and, conditions)

def cache_initiated():
    """Hackish function to test if there is an existing requests_cache"""
    try:
        requests_cache.get_cache()
        return True
    except AttributeError:
        return False


def get_decimal_encoder(prec=None, BaseEncoder=json.JSONEncoder):
    """Returns a JSONEncoder that forces Decimals to a given precision
    on string dump.

    Hackish worksround as these suggested solutions did not work in python3:
    https://stackoverflow.com/questions/1447287/format-floats-with-standard-json-module

    Example usage:

        data = {
            "value": Decimal(1.543),
        }
        print(json.dumps(data, cls=get_decimal_encoder(2)))
        # '{"value": 1.54}'

    :param prec (int): number of decimals
    """
    class DecimalEncoder(BaseEncoder):
        def default(self, o):
            if isinstance(o, Decimal):
                p = self.PRECISION
                if p is not None:
                    # Make the decimal a string with N decimal and then float it
                    o_str ='%.{0}f'.format(p) % o
                    return float(o_str)
                else:
                    return float(o)
            return super(DecimalEncoder, self).default(o)

    DecimalEncoder.PRECISION = prec

    return DecimalEncoder
